### PR TITLE
Refactor transformer decoder

### PR DESCRIPTION
**Pull request by aaprasad**
Refactor transformer implementation as requested by #34.

* emphasis on better variable names
* remove confused code paths
* store embeddings as a dictionary and always return that dictionary instead of conditionally returning None

## Summary by CodeRabbit

- **Refactor**
	- Improved type annotations in methods across transformer models for better code clarity and error checking.
- **Tests**
	- Enhanced clarity and consistency in variable naming within transformer model test cases.

### CODE REVIEW

File: biogtr/models/transformer.py [hunk 1]
```diff
@@ -179,8 +187,23 @@ def forward(self, frames: list[Frame], query_frame: int = None):
 
         emb = emb.permute(1, 0, 2)  # (total_instances, batch_size, embed_dim)
 
-        query_inds = None
+        batch_size, total_instances, embed_dim = reid_features.shape
+
+        reid_features = reid_features.permute(
+            1, 0, 2
+        )  # (total_instances x batch_size x embed_dim)
```
Comment (talmo) [inline biogtr/models/transformer.py:194]: Use standard shape tuple syntax for documenting shapes:

```suggestion
        )  # (total_instances, batch_size, embed_dim)
```

File: biogtr/models/transformer.py [hunk 2]
```diff
@@ -160,13 +164,17 @@ def forward(self, frames: list[Frame], query_frame: int = None):
...
         total_instances = sum(instances_per_frame)
         embed_dim = reid_features.shape[-1]
-
+        embeddings_dict = {"pos": None, "temp": None}
         # print(f'T: {window_length}; N: {total_instances}; N_t: {instances_per_frame} n_reid: {reid_features.shape}')
         pred_box, pred_time = get_boxes_times(frames)  # total_instances x 4
 
         temp_emb = self.temp_emb(pred_time / window_length)
+        if self.return_embedding:
+            embeddings_dict["temp"] = temp_emb  # .detach().cpu()
```
Comment (talmo) [inline biogtr/models/transformer.py:173]: unused comment

File: biogtr/models/transformer.py [hunk 3]
```diff
@@ -160,13 +164,17 @@ def forward(self, frames: list[Frame], query_frame: int = None):
...
 
         pos_emb = self.pos_emb(pred_box)
+        if self.return_embedding:
+            embeddings_dict["pos"] = pos_emb  # .detach().cpu()
```
Comment (talmo) [inline biogtr/models/transformer.py:177]: unused comment

File: biogtr/models/transformer.py [hunk 4]
```diff
@@ -179,8 +187,23 @@ def forward(self, frames: list[Frame], query_frame: int = None):
...
-        query_inds = None
+        batch_size, total_instances, embed_dim = reid_features.shape
+
+        reid_features = reid_features.permute(
+            1, 0, 2
+        )  # (total_instances x batch_size x embed_dim)
+
+        encoder_queries = reid_features
+
+        encoder_logits = self.encoder(
```
Comment (talmo) [inline biogtr/models/transformer.py:198]: Are you sure these are logits? That term is usually reserved for the pre-softmax/sigmoid activations in a classification head.

Maybe `encoder_features`?

File: biogtr/models/transformer.py [hunk 5]
```diff
@@ -367,32 +283,38 @@ def __init__(
...
+
+        queries = queries + pos_emb
+
+        # q = k = src
 
-        src2 = self.self_attn(
-            q,
-            k,
-            value=src,
+        attn_logits = self.self_attn(
```
Comment (talmo) [inline biogtr/models/transformer.py:305]: `attn_features`?

File: biogtr/models/transformer.py [hunk 6]
```diff
@@ -367,32 +283,38 @@ def __init__(
...
-        src = src + self.dropout2(src2)
-        src = self.norm2(src)
+        queries = queries + self.dropout1(attn_logits)
+        queries = self.norm1(queries)
+        projection = self.linear2(self.dropout(self.activation(self.linear1(queries))))
+        queries = queries + self.dropout2(projection)
+        encoder_logits = self.norm2(queries)
 
-        return src
+        return encoder_logits
```
Comment (talmo) [inline biogtr/models/transformer.py:317]: `encoder_features`?

**Review by talmo — APPROVED** [review]
Looks amazing as-is, but would make some small changes
